### DISCUSSION TITLE
p/0a559be724b14b28953dc7695f4d1246/

### DISCUSSION THREAD

**Discussion by giscus** [rxliuli/blog#30] [Announcements]
# p/0a559be724b14b28953dc7695f4d1246/

场景这两天由于 google 正在尝试一个垄断 web 浏览器的提案 Web Environment Integrity Explainer，所以吾辈打算再回归到 firefox 看看，下面记录一些遇到的问题和解决方法。 禁用在使用 devtools console 时提示 allow pasting每次想使用控制台时 firefox 都会提示，这很烦人。可以通过 about:config 禁用这

https://blog.rxliuli.com/p/0a559be724b14b28953dc7695f4d1246/

**Comment by yingziwu** [comment]
在RSS订阅中看到这篇博文，感到Orz。

也许是平台（主力 Arch Linux）差异，这篇博文中的好多问题我都没有遇见过。

> 禁用在使用 devtools console 时提示 allow pasting

手动输入 allow pasting 要求在我这只是每个 Profile 第一次粘贴时会提示，输过一次之后以后就再也不用输了。

> firefox 无法建立 websocket 连接

可能是平台差异，没有遇到过这样的问题。

这么重大的问题，也许可以到 bugzilla 报 bug。

> 如何修改 devtools 的界面为英文

同样好奇。

> firefox 删除标签页右键菜单无用的功能

![image](https://github.com/rxliuli/blog/assets/22741471/3fd1566e-67f5-4119-ba7b-f342920fa723)

Duplicate Tab、 Move Tab、Send Tab to Device、Open in New Container Tab 这几个功能我倒是经常使用的。

如何移除 tab menus 搜了一下，好像没有找到相关的内容。通过 Firefox CSS 也许可以做到。

不过，可以使用 [menus.create()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/create) 向 tab menus 中添加选项。

> firefox 有办法使用本地多用户么？

这个，Firefox  Multi Containers  并不是 Multi-Profile 替代品，更谈不上什么不完全替代吧。
Firefox  Multi Containers 我个人感觉很大程度上是为了 [Facebook Container](https://addons.mozilla.org/en-US/firefox/addon/facebook-container/) 才搞出来的。

Firefox  的 Profile  Manager 很早很早就有，相较 Chrome 只是没有在图形界面上提供一个 Other profiles 的按钮。

> 某些网站禁用了 firefox

Chrome 就是新 IE 诚不欺我。

像国内的某些网站，只能IE访问，而且还只能特定版本IE访问，这到底是不遵守Web标准，使用IE私有API的网站的问题，还是遵守Web标准的浏览器的问题？

> 某些字体并未遵循 css 或浏览器设定

特别看了一下 stylus 的源码，非常好奇论坛中提到的 Firefox bug 导致 stylus 不可用的Bug到底是什么。

![image](https://github.com/rxliuli/blog/assets/22741471/4195504c-c5ea-4a70-a366-51d1b1f21cee)
![image](https://github.com/rxliuli/blog/assets/22741471/a0709eed-cef7-4312-b685-e009d4b9692e)

可能的嫌疑犯？
[tabs.insertCSS()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/insertCSS)  使用了 ``matchAboutBlank`` 参数？
[webNavigation.onCommitted](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webNavigation/onCommitted) filter parameter 为空？

> firefox 禁止国内用户安装 ublock origin 广告屏蔽扩展

这个事件最让人不解的是，北京谋智火狐2018年便输了官司，但直到2022年3月（不早于 2022年3月8日，不晚于 2022年3月19日）才开始限制中国大陆IP访问部分广告屏蔽插件。

> Content-Security-Policy 默认配置导致某些浏览器插件不可用

非常好奇这是什么 Web Clipper 插件了。

Manifest V2 的 [tabs.executeScript()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/executeScript) 向 Page scripts 注入的代码在 Firefox 上需要遵守网站CSP，而在 Chromium 上无需遵守网站CSP，这一点老早就有人提出了，但到底是不是 Firefox 的 Bug 就见仁见智了（很明显 Firefox 开发人员不认为这是 Bug）。

但是注入到 Content scripts  无论是在 Firefox 还是在 Chromium 上均不会受到网站 CSP 的影响。而且在 Manifest V2 ，Content scripts 的 xhr 请求连 CORS 的限制也没有。
另外，与本地应用通迅这样的工作，不应该在 Background scripts 使用 [runtime.connectNative()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connectNative) API 进行吗？

> devtools network 不能看 request 和 response 选项卡

![image](https://github.com/rxliuli/blog/assets/22741471/5066ff9f-621f-4470-b708-e023ee2843b5)

刚刚试了一下，[1795291](https://bugzilla.mozilla.org/show_bug.cgi?id=1795291) 在我这里并没有能够复现。

User Agent : Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/116.0

--------------------------------------

我对 Firefox 的怨念主要是 API 的支持实在有一些慢。

- Manifest V3 直到 Firefox 109 才支持。（当然 Manifest V3 是谷歌搞出来的，所以这一条可以忽略）
- [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) 直到 111 才初步支持，而且至今仍不支持 [window.showOpenFilePicker()](https://developer.mozilla.org/en-US/docs/Web/API/window/showOpenFilePicker)、[window.showSaveFilePicker()](https://developer.mozilla.org/en-US/docs/Web/API/window/showSaveFilePicker)、[window.showDirectoryPicker()](https://developer.mozilla.org/en-US/docs/Web/API/window/showDirectoryPicker)。
- 直到 103 才完整支持 [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API)。

**Reply by rxliuli** [reply]
有些问题可能是吾辈遇到的特例（很难说，吾辈会认为这只是少见，不太相信吾辈会是第一个遇到这些问题的人），像是 discord 因为 ws 连接一直失败不能用或是 devtools network 中的 response 不正确显示的问题，但吾辈并不能找到问题的原因，也无法解决它。
某些问题则是通用的 dx 问题，像是 :has 或是 devtools console 的问题，以及部分网站（例如飞书）不兼容的问题，当然可以说这是网站的问题，chrome 是新 ie 什么的，但当它是必要的工具时，是谁的错误并不重要。就像 3q 大战中，普通用户真的没有选择。。。所以吾辈不得不滚回 chrome 了，也许最终会选择 vivaldi 之类的其他 chromium 内核浏览器
